PyNvCodec/TC/src/MemoryInterfaces.cpp

```diff
@@ -1006,3 +1020,257 @@ SurfaceYUV444::SurfaceYUV444(uint32_t width, uint32_t height, CUcontext context)
+}
+
+SurfaceRGB32FPlanar::~SurfaceRGB32FPlanar() = default;
+
+SurfaceRGB32FPlanar::SurfaceRGB32FPlanar() = default;
+
+SurfaceRGB32FPlanar::SurfaceRGB32FPlanar(const SurfaceRGB32FPlanar &other)
+    : plane(other.plane) {}
+
+SurfaceRGB32FPlanar::SurfaceRGB32FPlanar(uint32_t width, uint32_t height,
+                                   CUcontext context)
+    : plane(width, height * 3, ElemSize(), context) {}
+
+SurfaceRGB32FPlanar &SurfaceRGB32FPlanar::operator=(const SurfaceRGB32FPlanar &other) {
+  plane = other.plane;
+  return *this;
+}
+
+Surface *SurfaceRGB32FPlanar::Clone() { return new SurfaceRGB32FPlanar(*this); }
+
+Surface *SurfaceRGB32FPlanar::Create() { return new SurfaceRGB32FPlanar; }
+
+uint32_t SurfaceRGB32FPlanar::Width(uint32_t planeNumber) const {
+  if (planeNumber < NumPlanes()) {
+    return plane.Width();
+  }
+
+  throw invalid_argument("Invalid plane number");
+}
+
+uint32_t SurfaceRGB32FPlanar::WidthInBytes(uint32_t planeNumber) const {
+  if (planeNumber < NumPlanes()) {
+    return plane.Width() * plane.ElemSize();
+  }
+
+  throw invalid_argument("Invalid plane number");
+}
+
+uint32_t SurfaceRGB32FPlanar::Height(uint32_t planeNumber) const {
+  if (planeNumber < NumPlanes()) {
+    return plane.Height() / 3;
+  }
+
+  throw invalid_argument("Invalid plane number");
+}
+
+uint32_t SurfaceRGB32FPlanar::Pitch(uint32_t planeNumber) const {
+  if (planeNumber < NumPlanes()) {
+    return plane.Pitch();
+  }
+
+  throw invalid_argument("Invalid plane number");
+}
+
+uint32_t SurfaceRGB32FPlanar::HostMemSize() const {
+  return plane.GetHostMemSize();
+}
+
+CUdeviceptr SurfaceRGB32FPlanar::PlanePtr(uint32_t planeNumber) {
+  if (planeNumber < NumPlanes()) {
+    return plane.GpuMem() + planeNumber * Height() * plane.Pitch();
+  }
+
+  throw invalid_argument("Invalid plane number");
+}
+
+void SurfaceRGB32FPlanar::Update(const SurfacePlane &newPlane) {
+  plane = newPlane;
+}
+
+bool SurfaceRGB32FPlanar::Update(SurfacePlane *pPlanes, size_t planesNum) {
+  if (pPlanes && 1 == planesNum && !plane.OwnMemory()) {
+    plane = *pPlanes;
+    return true;
+  }
+
+  return false;
+}
+
+SurfacePlane *SurfaceRGB32FPlanar::GetSurfacePlane(uint32_t planeNumber) {
+  return planeNumber ? nullptr : &plane;
+}
+
+//=====================
+SurfaceRGB32FPlanarContiguous::~SurfaceRGB32FPlanarContiguous() = default;
+
+SurfaceRGB32FPlanarContiguous::SurfaceRGB32FPlanarContiguous() = default;
+
+SurfaceRGB32FPlanarContiguous::SurfaceRGB32FPlanarContiguous(
+    const SurfaceRGB32FPlanarContiguous &other)
+    : plane(other.plane), width(other.Width()), height(other.Height()), numPlanes(other.NumPlanes()) {}
+
+SurfaceRGB32FPlanarContiguous::SurfaceRGB32FPlanarContiguous(uint32_t _width,
+                                                             uint32_t _height,
+                                                             CUcontext context)
+    : plane(_width * _height * 3, 1, ElemSize(), context),
```

Comment:
How well does it work with 2D CUDA memcopies which are used to upload / download `Surface`?

Reply:
`DownloadSingleSurface` works, I have tested. 

I don't find any upload Surface function.

Following code demonstrates the utility of 1D image format.

``` python
    s_src = nvc.Surface.Make(nvc.PixelFormat.RGB_32F_PLANAR, 416, 416, 0)
    print(
        f"| Width: {s_src.Width()} Height: {s_src.Height()} Pitch: {s_src.Pitch()} NumPlanes: {s_src.NumPlanes()} HostSize: {s_src.HostSize()}"
    )

    s = nvc.Surface.Make(nvc.PixelFormat.RGB_32F_PLANAR_CONTIGUOUS, 416, 416,
                         0)
    print(
        f"| Width: {s.Width()} Height: {s.Height()} Pitch: {s.Pitch()} NumPlanes: {s.NumPlanes()} HostSize: {s.HostSize()}"
    )
    s_src.PlanePtr().Export(s.PlanePtr().GpuMem(), s.Pitch(), gpu_id)
    print(
        f"| Width: {s.Width()} Height: {s.Height()} Pitch: {s.Pitch()} NumPlanes: {s.NumPlanes()} HostSize: {s.HostSize()}"
    )
```

For resize functionalities I am including ResizeContext for Source ROI based resize. Will raise a seperate PR for that and would like to incorporate the 32F resize functionalities during that.

Reply:
>I don't find any upload Surface function.

It's done in `CudaUploadFrame` class.